Common/src/QuasarThreadPool.cpp

```diff
@@ -48,39 +48,94 @@ ThreadPool::~ThreadPool ()
     while (!m_pendingJobs.empty())
     {
         ThreadPoolJob* job = m_pendingJobs.front();
-        m_pendingJobs.pop();
+        m_pendingJobs.pop_front();
         LOG(Log::WRN) << "Removing unfinished job: " << job->describe();
         delete job;
     }
 }
 
+/** This method finds the next job suitable to be dealt by the next available worker.
+ * It can finish with two potential outcomes:
+ * -- there is no suitable job to execute 
+ *      returns nullptr for the job, does not change the job list
+ * -- there is a suitable job to execute (either w/o a mutex or with a mutex that is free)
+ *      locks that specifix mutex (if applicable)
+ *      returns the job ptr and the lock
+ *      removes that job from the list
+ */ 
+ThreadPool::Duty ThreadPool::findSomeDuty ()
+{
+    for (auto iter = std::begin(m_pendingJobs); iter != std::end(m_pendingJobs); iter++)
+    {
+        if (! (*iter)->associatedMutex() ) // no synchro domain
+        {
+            Duty duty;
+            duty.job = *iter;
+            m_pendingJobs.erase(iter);
+            LOG(Log::TRC) << "Removed job from the threadpool, current number of jobs is:" << m_pendingJobs.size();
+            return duty;
+        }
+        if ( m_mutices[(*iter)->associatedMutex()] == MutexUsage::OPEN ) // there is a synchro domain, but at least from the thread pool perspective it is not used.
+        {
+            Duty duty;
+            /* can we grab it ? */
+            std::unique_lock<std::mutex> lock (*(*iter)->associatedMutex(), std::try_to_lock);
+            if (!lock.owns_lock())
+                return duty; // no luck, someone else (outside of threadpool) owns it. next time.
+            /* so, we own the lock... */
+            m_mutices[(*iter)->associatedMutex()] = MutexUsage::CLOSED;
+            duty.lock = std::move(lock);
+            duty.job = *iter;
+            m_pendingJobs.erase(iter);
+            return duty;
+        }   
+    }
+    return Duty(); // by default no job, i.e. can't find anything to do now.
+}
+
+void ThreadPool::atNewCycle()
+{
+    for (auto& keyValPair : m_mutices)
+    {
+        if (keyValPair.second == MutexUsage::NEW_CYCLE)
+         keyValPair.second = MutexUsage::OPEN;
+    }
+}
+
 void ThreadPool::work()
 {
     while (!m_quit)
     {
-        std::unique_lock<std::mutex>lock (m_accessLock);
-        if (! m_pendingJobs.empty())
+        Duty duty;
```

Comment:
Consider using std::optional if a duty does not match a job or unique_ptr of the job in the Duty struct.